daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala

```diff
@@ -74,6 +76,21 @@ abstract class UState(
   extends ParseOrUnparseState(vbox, diagnosticsArg, dataProcArg, tunable)
   with Cursor[InfosetAccessor] with ThrowsSDE with SavesErrorsAndWarnings {
 
+  final def setVariable(vrd: VariableRuntimeData, newValue: DataValuePrimitive, referringContext: ThrowsSDE) =
+    vbox.vmap.setVariable(vrd, newValue, referringContext, this)
+
+  /**
+   * For unparsing, this throws a RetryableException in the case where the variable cannot (yet) be read.
+   *
+   * @param vrd Identifies the variable to read.
+   * @param referringContext Where to place blame if there is an error.
+   * @return The data value of the variable, or throws exceptions if there is no value.
+   */
+  final def getVariable(vrd: VariableRuntimeData, referringContext: ThrowsSDE): DataValuePrimitive =
```

Comment:
Same comment as above, should we be using these set/getVariable functions internally? I assume right now we are just directly access this variable map?

Reply:
I will actually look at how expressions that read variables do it. They are obviously not calling this getter, but perhaps could which would solve the test coverage.

Reply:
Created DAFFODIL-2565 for this issue. VRef in DPath was not maintaining the changedVariablesStack.

Reply:
Turns out, there was no bug here. VariableMap.readVariable was, in fact, doing a callback to PState to maintain the changedVariablesStack, so this code was broken because it was adding variables twice. 

I changed DAFFODIL-2565 into an improvement cleanup, as it is cleaner if all variable manipulation happens via the PState. Ideally the variableMap would be a private pstate member, but the debugger currently has access to it.